source/MaterialXRuntime/Commands/PrimCommands.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+void RtRenamePrimCmd::execute(RtCommandResult& result)
+{
+    try
+    {
+        RtToken oldName = _path.getName();
+        RtToken resultName = _stage->renamePrim(_path, _newName);
+
+        // Update the path and name so we can undo later
+        _path.pop();
+        _path.push(resultName);
+        _newName = oldName;
+
+        result = RtCommandResult(true);
+    }
+    catch (const ExceptionRuntimeError& e)
+    {
+        result = RtCommandResult(false, string(e.what()));
+    }
+}
+
+void RtRenamePrimCmd::undo(RtCommandResult& result)
+{
+    try
+    {
+        // Path and name was updated after execution
+        // so we can just re-execute the command.
+        execute(result);
+
+        result = RtCommandResult(true);
+    }
+    catch (const ExceptionRuntimeError& e)
+    {
+        result = RtCommandResult(false, string(e.what()));
+    }
+}
+
+
+RtCommandPtr RtReparentPrimCmd::create(RtStagePtr stage, const RtPath& path, const RtPath& newParentPath)
+{
+    return std::make_shared<RtReparentPrimCmd>(stage, path, newParentPath);
+}
+
+void RtReparentPrimCmd::execute(RtCommandResult& result)
+{
+    try
+    {
+        RtToken resultName = _stage->reparentPrim(_path, _parentPath);
+
+        // Update paths so we can undo later
+        RtPath newPath = _parentPath;
+        newPath.push(resultName);
+        _parentPath = _path;
+        _parentPath.pop();
+        _path = newPath;
+
+        result = RtCommandResult(true);
+    }
+    catch (const ExceptionRuntimeError& e)
+    {
+        result = RtCommandResult(false, string(e.what()));
+    }
+}
+
+void RtReparentPrimCmd::undo(RtCommandResult& result)
+{
+    try
+    {
+        RtToken resultName = _stage->reparentPrim(_path, _parentPath);
+
+        // Make sure we get back to the original name
+        // in case the reparenting forced the prim name to change.
+        if (resultName != _originalName)
+        {
+            RtPath newPath = _parentPath;
+            newPath.push(resultName);
+            _stage->renamePrim(newPath, _originalName);
+            resultName = _originalName;
+        }
+
+        // Update paths so we can redo later
+        RtPath newPath = _parentPath;
+        newPath.push(resultName);
+        _parentPath = _path;
+        _parentPath.pop();
+        _path = newPath;
+
+        result = RtCommandResult(true);
+    }
+    catch (const ExceptionRuntimeError& e)
+    {
+        result = RtCommandResult(false, string(e.what()));
+    }
+}
+
+namespace RtCommand
+{
+
+    void createPrim(RtStagePtr stage, const RtToken& typeName, RtCommandResult& result)
+    {
+        RtCommandPtr cmd = RtCreatePrimCmd::create(stage, typeName, RtPath("/"), EMPTY_TOKEN);
+        RtApi::get().getCommandEngine().execute(cmd, result);
+    }
+
+    void createPrim(RtStagePtr stage, const RtToken& typeName, const RtPath& path, RtCommandResult& result)
+    {
+        RtPath parentPath(path);
+        parentPath.pop();
```

Comment:
Just wondering here. What is the parent of a top level primitive, or is this just the root path returned?
Does it make sense to require a parent path to be provided instead?

Reply:
Parent of a top level prim is the stage root prim. There is another variation of this command below where you give the parent path and prim name instead of full path.